### PR TITLE
Add textcolor to legend based on labelcolor string

### PR DESCRIPTION
**Pull request by chahak13**
## PR Summary

As raised by #20577, setting `labelcolor` to any of the string options did not work with a scatter plot as it is a PathCollection with possible multiple color values. This commit fixes that. Now, if there is a colormap or a scatter plot with multiple values, then, the legend text color is not changed, but otherwise, the text color is changed to the color of the scatter markers.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by chahak13** [comment]
@timhoffm as we discussed, I started working on issue #20577. I wanted some suggestions if you could help. Did not want to clutter the issue discussion with code questions so moved to this working PR.

My idea is that if we have a `LinearSegmentedColormap`, then we should not be changing the text color. If not, we check if we have a single color/same color in an array and proceed accordingly. These seemed to work on at least the simple cases but I ran into one issue while writing the tests for this. If I have some scatter plots on an axes, like

```python
fig, ax = plt.subplots()
ax.scatter(np.arange(10), np.arange(10)*1, label='#1', c='r')
ax.scatter(np.arange(10), np.arange(10)*2, label='#2', c='g')
ax.scatter(np.arange(10), np.arange(10)*3, label='#3', c='b')
leg = ax.legend(labelcolor="linecolor")
```
This creates a `ListedColormap` which is what I was expecting, but when I write a function in `test_legend.py` as 


```python
def test_legend_pathcollection_labelcolor_linecolor():
    fig, ax = plt.subplots()
    ax.scatter(np.arange(10), np.arange(10)*1, label='#1', c='r')
    ax.scatter(np.arange(10), np.arange(10)*2, label='#2', c='g')
    ax.scatter(np.arange(10), np.arange(10)*3, label='#3', c='b')

    leg = ax.legend(labelcolor='linecolor')
    for text, color in zip(leg.get_texts(), ['r', 'g', 'b']):
        assert mpl.colors.same_color(text.get_color(), color)
```
the generated cmap comes out to be `LinearSegmentedColormap` (Found using `handle.get_cmap()` in both cases). Why would there be a different behavior in both cases? Or am I doing something wrong? Thanks!

**Comment by timhoffm** [comment]
I admit this is quite arcane for historic reasons. All collections have a colormap, even if they don't use it.
What you need to check is whether the data values [get_array()](https://matplotlib.org/devdocs/api/collections_api.html#matplotlib.collections.PathCollection.get_array) are `None`.

**Comment by chahak13** [comment]
@timhoffm Thanks for that! I've updated and added tests to the PR. I was not sure where to document this change though. Let me know what else I need to do?

**Review by timhoffm — APPROVED** [review]
This is good to go modulo some optional improvements in the tests.

@chahak13 if we consider this only a bugfix, there's no additional documentation needed.

**Comment by chahak13** [comment]
@timhoffm thanks! I updated the tests. Should be good now :)

File: lib/matplotlib/tests/test_legend.py [hunk 1]
```diff
@@ -671,6 +671,41 @@ def test_legend_labelcolor_linecolor():
...
+def test_legend_pathcollection_labelcolor_linecolor_iterable():
+    # test the labelcolor for labelcolor='linecolor' on PathCollection
+    # with iterable colors
+    fig, ax = plt.subplots()
+    colors = np.random.default_rng().choice(['r', 'g', 'b'], 10)
+    ax.scatter(np.arange(10), np.arange(10)*1, label='#1', c=colors)
+
+    leg = ax.legend(labelcolor='linecolor')
+    text, = leg.get_texts()
+    assert mpl.colors.same_color(text, 'black')
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_legend.py:695]:
```suggestion
    assert mpl.colors.same_color(text.get_text(), 'black')
```

File: lib/matplotlib/tests/test_legend.py [hunk 2]
```diff
@@ -671,6 +671,41 @@ def test_legend_labelcolor_linecolor():
...
+
+def test_legend_pathcollection_labelcolor_linecolor_cmap():
+    # test the labelcolor for labelcolor='linecolor' on PathCollection
+    # with a colormap
+    fig, ax = plt.subplots()
+    ax.scatter(np.arange(10), np.arange(10), c=np.arange(10), label='#1')
+
+    leg = ax.legend(labelcolor='linecolor')
+    text, = leg.get_texts()
+    assert mpl.colors.same_color(text, 'black')
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_legend.py:706]:
```suggestion
    assert mpl.colors.same_color(text.get_text(), 'black')
```

**Review by timhoffm — COMMENTED** [review]
Whoops, sorry my suggestion code was incorrect. This should fix CI.

**Comment by chahak13** [comment]
Right, sorry. Should have spotted that too. Thanks!

**Comment by chahak13** [comment]
@timhoffm changed it to the correct `get_color` call. Tests should all be fine now. Let me know if there's anything else. Thanks!

**Review by timhoffm — APPROVED** [review]

**Comment by tacaswell** [comment]
Thanks @chahak13 and congratulations on your first merged Matplotlib PR 🎉 (I know you have at least one   more in flight).  I hope we continue to hear from you.


------

Even though this is a bug fix, it is a fair bit of code and I do not feel comfortable backporting this no 3.6.x.

**Pull request merged**